### ISSUE TITLE
Improve automatic creation of myLeoSettings.leo

### ISSUE THREAD

**Issue by ghost** [leo-editor/leo-editor#282]
By default, Leo seems to be installed with a settings node called @enabled-plugins in myLeoSettings.leo that is identical to the same node in leoSettings.leo.

In this @enable-plugins node, instructions say that in order to enable plugins other than those already enabled in leoSettings.leo, one should modify the @enabled-plugins node in myLeoSettings.leo.

However, the following inconsistent behavior can be observed:
- if one deletes the @enabled-plugins node in myLeoSettings.leo altogether, all plugins enabled in leoSettings.leo are loaded when restarting, as they should;
- if one deletes all text in the @enabled-plugins node in myLeoSettings.leo but keep the empty node, no plugins are loaded by Leo at all when restarting, no matter what plugins are enabled in leoSettings.leo;
- if one deletes all text in the body of the @enabled-plugins node in myLeoSettings.leo, and then write a few entries for some plugin to be enabled in the node body, no plugins are loaded by Leo at all when restarting, no matter what plugins are enabled in leoSettings.leo;
- if one only leaves the first lines, from "# Leo loads plugins in ..." to  "# Alphabetical list of all Leo plugins." in the body of the @enabled-plugins node in myLeoSettings.leo, but deletes all remaining text in the node body and then write a few entries for some plugin to be enabled, all plugins enabled in leoSettings.leo are loaded when restarting.

The above is inconsistent and counter-intuitive: since settings in myLeoSettings.leo are supposed to override the corresponding ones in leoSettings.leo, one would expect that leaving a blank (= empty body) @enabled-settings node in myLeoSettings.leo would not affect whether the plugins enabled in leoSettings.leo are loaded or not.

Similarly, one would expect that an @enabled-plugins node in myLeoSettings.leo only containing, say, one line to enable one additional plugin, should not affect whether the plugins enabled in leoSettings are loaded or not: the only effect should be that the additional plugin is enabled and loaded.

The behavior of these settings should be made more consistent and intuitive, or at least it should be noted in the instructions that there is a mandatory part  (the first lines as explained above) that must be present/not deleted in the @enabled-plugins node body in myLeoSettings.leo in order for the plugins to be loaded correctly.

**Comment by edreamleo** [comment]
> By default, Leo seems to be installed with a settings node called @enabled-plugins in myLeoSettings.leo that is identical to the same node in leoSettings.leo.

No.  Leo is installed without any myLeoSettings.leo.  If no myLeoSettings.leo file is found, Leo searches for settings in the local .leo file, and then in leoSettings.leo. This is explained as clearly as possible [here](http://leoeditor.com/customizing.html#search-order-for-settings-files).

Furthermore, the @enabled-plugin setting could be called "cumulative", because once a plugin is enabled there is no way to "un-enable" a plugin.

In future, please ask for help at [Leo's Google Forum](https://groups.google.com/forum/#!forum/leo-editor) before posting bug reports.

**Comment by tbnorth** [comment]
On Mon, 02 May 2016 06:08:23 -0700
"Edward K. Ream" <notifications@github.com> wrote:

> > By default, Leo seems to be installed with a settings node called
> > @enabled-plugins in myLeoSettings.leo that is identical to the same
> > node in leoSettings.leo.
>
> No.  Leo is installed without any myLeoSettings.leo.  If no
> myLeoSettings.leo file is found, Leo searches for settings in the
> local .leo file, and then in leoSettings.leo. This is explained as
> clearly as possible
> [here](http://leoeditor.com/customizing.html#search-order-for-settings-files).

The docs. might need tweaking to mention that while the above describes
how Leo searches for existing settings, if you attempt to open (through
the Settings menu) myLeoSettings.leo when it doesn't exist, it's
automagically created with a copy of @enabled-plugins from
leoSettings.leo.  The text in the visible node of the new
myLeoSettings.leo mentions that it was created "20xx mmm dd hh:mm" which
will be seconds before, but other than that, it might not be obvious
that myLeoSettings.leo didn't exist until you asked for it.

Cheers -Terry

**Comment by edreamleo** [comment]
On Mon, May 2, 2016 at 9:24 AM, Terry Brown <notifications@github.com>
wrote:

>
> The docs. might need tweaking to mention that while the above describes
> how Leo searches for existing settings, if you attempt to open (through
> the Settings menu) myLeoSettings.leo when it doesn't exist, it's
> automagically created with a copy of @enabled-plugins from
> leoSettings.leo.


​Thanks for this.  Will do.

EKR

**Comment by edreamleo** [comment]
On Mon, May 2, 2016 at 9:27 AM, Edward K. Ream <edreamleo@gmail.com> wrote:

> On Mon, May 2, 2016 at 9:24 AM, Terry Brown <notifications@github.com>
> wrote:
>
>>
>> The docs. might need tweaking to mention that while the above describes
>> how Leo searches for existing settings, if you attempt to open (through
>> the Settings menu) myLeoSettings.leo when it doesn't exist, it's
>> automagically created with a copy of @enabled-plugins from
>> leoSettings.leo.
>
>
> ​Thanks for this.  Will do.
>

​Hmm.  On second thought, I wonder whether the automagic is really a good
idea.  In the present instance it seems to have caused confusion.  And is
copying *all* of leoSettings.leo really a good idea?  And where should Leo
create myLeoSettings.leo?

Edward

**Comment by tbnorth** [comment]
On Mon, 02 May 2016 11:51:05 -0700
"Edward K. Ream" <notifications@github.com> wrote:

> On Mon, May 2, 2016 at 9:27 AM, Edward K. Ream <edreamleo@gmail.com>
> wrote:
> 
> > On Mon, May 2, 2016 at 9:24 AM, Terry Brown
> > <notifications@github.com> wrote:
> >
> >> The docs. might need tweaking to mention that while the above
> >> describes how Leo searches for existing settings, if you attempt
> >> to open (through the Settings menu) myLeoSettings.leo when it
> >> doesn't exist, it's automagically created with a copy of
> >> @enabled-plugins from leoSettings.leo.
> >
> > ​Thanks for this.  Will do.
> 
> ​Hmm.  On second thought, I wonder whether the automagic is really a
> good idea.  In the present instance it seems to have caused
> confusion.  And is copying *all* of leoSettings.leo really a good
> idea?  And where should Leo create myLeoSettings.leo?

I think automatically creating the file is worth keeping.  We could add
a popup to tell the user it was just created if we wanted it to be more
obvious.

Without automagic creation, the new user must correctly layout a Leo
settings file.

  @settings
    @enabled-plugins
    @shortcuts

etc. before they can change any settings, which I think is too much of
a barrier.

It does not copy all of leoSettings.leo, which would be a bad
idea.  It just copies key bindings and enabled-plugins.

It should be created in ~/.leo/myLeoSettings.leo, that's worked fine so
far.

Cheers -Terry

**Comment by edreamleo** [comment]
I am going to reopen this so that we can deal with the user interface questions.

**Comment by edreamleo** [comment]
I have just changed the title of this issue. I plan to follow Terry's advice to the letter.

**Comment by edreamleo** [comment]
Fixed at c68fabeb52.

I updated the docs.  More importantly, Leo now puts up a yes/no dialog asking the user whether they want to create myLeoSettings.leo in the particular directory. This allows the user to cancel the automatic creation of the file.

As I was testing this, I see that myLeoSettings.leo is actually created in a new (dirty) Leo tab, so the user can decline to save the created myLeoSettings.leo if the user doesn't like the look of the file.

I think this is enough for now.  I'm going to close this item.
